bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java

```diff
@@ -421,94 +436,187 @@ public LayoutManager getLayoutManager(){
+        }
+        // create available bookies node if not exists
+        if (!availableNodeExists) {
+            zk.create(conf.getZkAvailableBookiesPath(), "".getBytes(Charsets.UTF_8), zkAcls, CreateMode.PERSISTENT);
+        }
+
+        // create readonly bookies node if not exists
+        if (null == zk.exists(conf.getZkAvailableBookiesPath() + "/" + READONLY, false)) {
+            zk.create(conf.getZkAvailableBookiesPath() + "/" + READONLY, new byte[0], zkAcls, CreateMode.PERSISTENT);
+        }
+
+        return ledgerRootExists;
+    }
+
+    @Override
+    public boolean initNewCluster() throws Exception {
+        String zkLedgersRootPath = conf.getZkLedgersRootPath();
+        String zkServers = conf.getZkServers();
+        String zkAvailableBookiesPath = conf.getZkAvailableBookiesPath();
+        log.info("Initializing ZooKeeper metadata for new cluster, ZKServers: {} ledger root path: {}", zkServers,
+                zkLedgersRootPath);
+
+        boolean ledgerRootExists = null != zk.exists(conf.getZkLedgersRootPath(), false);
+
+        if (ledgerRootExists) {
+            log.error("Ledger root path: {} already exists", conf.getZkLedgersRootPath());
+            return false;
+        }
+
+        // Create ledgers root node
+        zk.create(zkLedgersRootPath, "".getBytes(UTF_8), Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+        // create available bookies node
+        zk.create(zkAvailableBookiesPath, "".getBytes(UTF_8), Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+        // creates the new layout and stores in zookeeper
+        LedgerManagerFactory.newLedgerManagerFactory(conf, layoutManager);
+
+        // create INSTANCEID
+        String instanceId = UUID.randomUUID().toString();
+        zk.create(conf.getZkLedgersRootPath() + "/" + BookKeeperConstants.INSTANCEID, instanceId.getBytes(UTF_8),
+                Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+        log.info("Successfully initiated cluster. ZKServers: {} ledger root path: {} instanceId: {}", zkServers,
+                zkLedgersRootPath, instanceId);
+        return true;
+    }
+
+
+    @Override
+    public boolean nukeExistingCluster() throws Exception {
+        String zkLedgersRootPath = conf.getZkLedgersRootPath();
+        String zkServers = conf.getZkServers();
+        log.info("Nuking ZooKeeper metadata of existing cluster, ZKServers: {} ledger root path: {}",
+                zkServers, zkLedgersRootPath);
+
+        boolean ledgerRootExists = null != zk.exists(conf.getZkLedgersRootPath(), false);
+        if (!ledgerRootExists) {
+            log.info("There is no existing cluster with ledgersRootPath: {} in ZKServers: {}, "
+                    + "so exiting nuke operation", zkLedgersRootPath, conf.getZkServers());
+            return true;
+        }
+
+        String availableBookiesPath = conf.getZkAvailableBookiesPath();
```

Comment:
The logic here is way too complicated. You don't need to read the list the znodes and check if they are still exists. You can just issue a zk txn (don't remember that exactly method, but the following code is trying to demo how it works), if there are still bookies alive, the txn will fail.

```
Multi multi = zk.multi();
multi.delete(zkLedgersRootPath + "/available/readonly")
multi.delete(zkLedgersRootPath + "/available");

multi.execute();
```

Reply:
if you do as what I suggested, you don't need to change BookieWatcher.

Reply:
Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue. Also, I moved deletions of the ZK nodes to  LedgerManagerFactory class, for the sake better abstraction. Anyhow the changes I made to BookieWatcher are minimal. For now, I feel comfortable to make independent zk operations, because check/delete operations are made at multiple classes with proper validation.

Reply:
> Here before proceeding with nuking cluster data, I would like to make sure that there are no connected bookies to the cluster and print informative log message for each step/issue.

the operation that I proposed is doing exactly same purpose as you hope, with much simpler approach and less error prone.

if you really feel the need to read and delete, you should not use BookieWatch, you should just use registration client api, rather than api in bookie watch. because bookie watch is the user of metadata interfaces, the metadata interface implementation should not use those api, this is a cycle dependency. 

These are the apis you should use: 

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L75
https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/RegistrationClient.java#L82

Reply:
reverting BookieWatcher changes and going to use RegistrationClient API